package com.fpf.blucon.ui.components.common

import androidx.compose.animation.Crossfade
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.text.input.OutputTransformation
import androidx.compose.foundation.text.input.TextFieldLineLimits
import androidx.compose.foundation.text.input.TextFieldState
import androidx.compose.foundation.text.input.clearText
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay
import kotlin.time.Duration.Companion.milliseconds

@OptIn(ExperimentalFoundationApi::class)
@Composable
fun SearchBar(
    searchFieldState: TextFieldState,
    enabled: Boolean,
    placeholders: List<String>,
    onSearch: () -> Unit,
    modifier: Modifier = Modifier,
    autoFocus: Boolean = false,
    placeholderChangeDuration: Long = 2000L,
    trailingIcon: @Composable (() -> Unit)? = null,
    leadingIcon: @Composable (() -> Unit)? = null,
    onFocusedChange: ((focused: Boolean) -> Unit)? = null
) {
    var currentPlaceHolder by remember { mutableStateOf(placeholders[0]) }

    val interactionSource = remember { MutableInteractionSource() }
    val isFocused by interactionSource.collectIsFocusedAsState()
    val outlineColor = if(isFocused) MaterialTheme.colorScheme.primary else Color.Transparent
    val tagColor =  MaterialTheme.colorScheme.onSurface.copy(0.5f)
    val ot = OutputTransformation {
        val regex = Regex("^#\\w+")
        regex.findAll(searchFieldState.text).forEach {
            match ->
            addStyle( SpanStyle(color = tagColor, fontStyle = FontStyle.Italic), match.range.first, match.range.last + 1 )
        }
    }


    val focusRequester = remember { FocusRequester() }

    LaunchedEffect(autoFocus) {
        if (autoFocus) focusRequester.requestFocus()
    }

    LaunchedEffect(isFocused) {
        onFocusedChange?.invoke(isFocused)
    }


    LaunchedEffect(placeholders, searchFieldState.text, isFocused) {
        if (isFocused || searchFieldState.text.isNotBlank() || currentPlaceHolder !in placeholders) {
            currentPlaceHolder = placeholders[0]
            return@LaunchedEffect
        }

        while (true) {
            delay(placeholderChangeDuration.milliseconds)
            val currentIdx = placeholders.indexOf(currentPlaceHolder)
            val nextIdx = if (currentIdx < placeholders.size - 1) currentIdx + 1 else 0
            currentPlaceHolder = placeholders[nextIdx]
        }
    }


    Box(
        modifier = modifier
            .fillMaxWidth()
            .heightIn(min = 56.dp)
            .border(width = 1.dp, color = outlineColor, shape = MaterialTheme.shapes.extraLarge)
            .clip(MaterialTheme.shapes.extraLarge)
//            .background(MaterialTheme.colorScheme.surfaceContainer)
    ) {
        BasicTextField(
            state = searchFieldState,
            interactionSource = interactionSource,
            enabled = enabled,
            lineLimits = TextFieldLineLimits.SingleLine,
            outputTransformation = ot,
            textStyle = MaterialTheme.typography.bodyLarge.copy(color = MaterialTheme.colorScheme.onSurface),
            keyboardOptions = KeyboardOptions.Default.copy(imeAction = ImeAction.Search),
            onKeyboardAction = {
                if (searchFieldState.text.isNotBlank()) {
                    onSearch()
                } else {
                    it()
                }
            },
            cursorBrush = SolidColor(MaterialTheme.colorScheme.onSurface),
            modifier = Modifier
                .fillMaxWidth()
                .focusRequester(focusRequester),
            decorator = { innerTextField ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(2.dp),
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 8.dp)
                        .heightIn(min = 56.dp)
                ) {
                    leadingIcon?.invoke()

                    Box(
                        contentAlignment = Alignment.CenterStart,
                        modifier = Modifier
                            .weight(1f)
                            .padding(vertical = 16.dp)
                    ) {
                        innerTextField()
                        if (searchFieldState.text.isEmpty()) {
                            Crossfade(targetState = currentPlaceHolder) { text ->
                                Text(
                                    text = text,
                                    style = MaterialTheme.typography.bodyLarge.copy(
                                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f)
                                    ),
                                )
                            }
                        }
                    }

                    if (searchFieldState.text.isNotBlank()) {
                        IconButton(
                            enabled = enabled,
                            onClick = { searchFieldState.clearText() },
                            modifier = Modifier
                                .align(Alignment.Top)
                                .padding(top = 4.dp)
                        ) {
                            Icon(
                                imageVector = Icons.Filled.Close,
                                contentDescription = "Clear query",
                                tint = MaterialTheme.colorScheme.surface,
                                modifier = Modifier
                                    .background(
                                        MaterialTheme.colorScheme.onBackground.copy(alpha = 0.5f),
                                        shape = CircleShape
                                    )
                                    .size(16.dp)
                                    .padding(2.dp)
                            )
                        }
                    }
                    Box(modifier = Modifier
                        .align(Alignment.Top)
                        .padding(top = 4.dp)){
                        trailingIcon?.invoke()
                    }
                }
            }
        )
    }
}
